common/src/main/java/io/netty/util/concurrent/SingleThreadEventExecutor.java

```diff
@@ -419,7 +419,8 @@ protected void wakeup(boolean inEventLoop) {
 
     @Override
     public boolean inEventLoop(Thread thread) {
-        return thread == this.thread;
+        return thread == this.thread ||
+                executor instanceof EventExecutor && ((EventExecutor) executor).inEventLoop(thread);
```

Comment:
This seems a bit odd. You mentioned earlier about getting deadlocks but I don't know why.

Couldn't we make ImmediateEventExecutor non final and then subclass it to make it implement EventLoop? That would eliminate some of the new ImmediateXXXPromise types here?

I may just be missing something fundamental here.